docs/plugins.md

```diff
@@ -14,183 +14,157 @@ Looking for help using Snowpack in your project?
-  </li>
-  <li class="tc-blue">
-    <h4 class="ol-title">Transform</h4>
-    <div class="tc-text">This stage happens after the <strong>build</strong> step, and can be used for additional transformations if needed. This has more context since it comes later, such as which URL the module was requested at.</div>
-  </li>
-  <li class="tc-magenta">
-    <h4 class="ol-title">Bundle</h4>
-    <div class="tc-text">Optional stage that only happens when using a bundler plugin such as <a href="https://github.com/pikapkg/create-snowpack-app/tree/master/packages/plugin-webpack" target="_blank" rel="noopener nofollow">@snowpack/plugin-webpack</a> or <a href="https://github.com/pikapkg/create-snowpack-app/tree/master/packages/plugin-parcel" target="_blank" rel="noopener nofollow">@snowpack/plugin-parcel</a>. This passes all of Snowpack’s built files to a final bundler. <em>Note: only 1 bundler plugin can receive the final files.</em></div>
-  </li>
-</ol>
+## Build Plugins
 
+Build plugins run on single files that pass through Snowpack. They listen for particular file extensions, and when a match is found, transform that code.
 
-## Plugin API
+In order to keep operations speedy, Snowpack’s build system is **single pass,** meaning each file is only transformed by a set of plugins once. We’ll talk more about that later.
 
-A Snowpack plugin is a JavaScript module that exports a function. That function takes Snowpack config & plugin options as it's function arguments and must return a valid Snowpack Plugin object.
-
-### Example
+### The Basics
 
 ```js
-// snowpack.config.json
-"plugins": [["@prefresh/snowpack", { /* `pluginOptions` (optional) */ }]]
+// Snowpack plugin template
+
+module.exports = (snowpackConfig, pluginOptions) => ({
+  name: 'my-snowpack-plugin', // plugin name
+  input: ['.js', '.ts'], // extensions to listen for
+  output: ['.js'], // extensions that will be output
+  async build(buildOptions) { /* … */ } // code transformations
+})
 ```
 
-```js
-// "@prefresh/snowpack": This plugin adds automatic HMR for Preact applications.
-module.exports = function createPlugin(snowpackConfig, pluginOptions) {
-  return {
-    knownEntrypoints: ['@prefresh/core'],
-    /**
-     * transform() - Transform web assets (JS, CSS, or HTML). Useful for
-     * post-processing or adding functionality into your web app.
-     */
-    async transform({ contents, urlPath, isDev }) {
-      if (!isDev) {
-        return;
-      }
-      if (!urlPath.endsWith('.js')) {
-        return;
-      }
-      return {
-        result: `
-          import '@prefresh/core';
-          ${contents}
-
-          if (import.meta.hot) {
-            /* IMPLEMENTATION, REMOVED FOR BREVITY */
-          }`,
-      };
-    },
-  };
-};
-```
+A build plugin is a **function** that accepts 2 parameters, in this order:
```

Comment:
interesting naming/terminology question: is the build plugin the function, or is it the object that is returned? In my mind it's always been the returned object (this is the model that Rollup uses) and the function is just a way to construct one.

Since the the stateful object is treated as the SnowpackPlugin in our codebase, I'd vote to match that in our documentation as well.

Reply:
Good question. I feel pretty strongly that it is the function. Since plugins can’t just export an object; they always have to export a function. But we do call the object a “plugin” internally since that’s what we work with.

Reply:
Cool. Updated the other PR with that discussion.